### PR TITLE
DAOS-4623 reintegrate: use cont key to check the container destory.

### PR DESCRIPTION
**Pull request by wangdi1**
Add cont_key, instead of rlink, into the cont destroy
hash.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

**Comment by wangdi1** [comment]
> This looks fine. Is D_LIST_INIT_HEAD needed on the rlink field for the hash table to work?
> 
> Storing the uuid that's already the key in the value again seems bizarre to me - the value is never used, so you shouldn't need to store anything there... This is still more clear than my initial implementation, but I don't understand why it'd fix any problems.

The initial implementation does not record the cont uuid into the hash, only insert a rlink, which make the hash useless. i.e. it will keep destroying the container during reintegration once the cont cb is called.

**Comment by byron-marohn** [comment]
But the key is the cont uuid? Isn't this "recording the cont uuid into the hash" ?  Why does the key also need to be stored in the hash table value? None of the other hash table library implementations I've used have worked this way.

File: src/object/srv_obj_migrate.c [hunk 1]
```diff
@@ -1418,14 +1418,13 @@ destroy_existing_container(struct migrate_pool_tls *tls, uuid_t cont_uuid)
 		 * to test set membership. The link stored is just the simplest
 		 * base list type
 		 */
-		d_list_t *rlink;
-
+		struct migrate_init_cont_key *key;
+
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/object/srv_obj_migrate.c:1422]: can you fix this one?

File: src/object/srv_obj_migrate.c [hunk 2]
```diff
@@ -1438,16 +1437,18 @@ destroy_existing_container(struct migrate_pool_tls *tls, uuid_t cont_uuid)
...
+		uuid_copy(key->cont_uuid, cont_uuid);
+		D_INIT_LIST_HEAD(&key->cont_link);
 		rc = d_hash_rec_insert(&tls->mpt_cont_dest_tab, cont_uuid,
-				       sizeof(uuid_t), rlink, true);
+				       sizeof(uuid_t), &key->cont_link, true);
 		if (rc) {
 			D_ERROR("Failed to insert uuid table entry "DF_RC"\n",
 				DP_RC(rc));
-			D_FREE(rlink);
+			D_FREE_PTR(key);
```
Comment (jolivier23) [inline src/object/srv_obj_migrate.c:1451]: D_FREE_PTR is deprecated (we've been trying to get rid of it).   D_FREE is sufficient.

**Comment by wangdi1** [comment]
> But the key is the cont uuid? Isn't this "recording the cont uuid into the hash" ? Why does the key also need to be stored in the hash table value? None of the other hash table library implementations I've used have worked this way.
Hmm, you have record the cont uuid into the hash, otherwise the compare will not use right uuid to compare.

**Pull request merged**